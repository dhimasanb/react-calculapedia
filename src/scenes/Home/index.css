.App {
  font-family: sans-serif;
  text-align: center;
}

.todoContainer {
  width: 50%;
  margin-left: auto;
  margin-right: auto;
}


/* .gx-login-container {
  position: relative;
  height: 100%;
  padding-bottom: 30px;
  -webkit-flex: 1;
  flex: 1 1;
  display: -webkit-flex;
  display: -ms-flex;
  display: flex;
  -webkit-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -ms-align-items: center;
  align-items: center;
  -webkit-justify-content: center;
  -ms-justify-content: center;
  justify-content: center;
}

.gx-login-content {
  background-color: #ffffff;
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  border-radius: 10px;
  -webkit-box-shadow: 0 0 5px 5px rgba(0, 0, 0, 0.03);
  -moz-box-shadow: 0 0 5px 5px rgba(0, 0, 0, 0.03);
  box-shadow: 0 0 5px 5px rgba(0, 0, 0, 0.03);
}

.gx-text-center {
  text-align: center!important;
}

.gx-login-header {
  margin-bottom: 30px;
}

.ant-avatar-image {
  background: transparent;
}

.ant-avatar {
  color: #545454;
  color: #fff;
  background: #ccc;
  border-radius: 50%;
}

.gx-size-120 {
  height: 120px!important;
  width: 120px!important;
  line-height: 120px!important;
}

.gx-mb-4, .gx-my-4 {
  margin-bottom: 1.5rem!important;
}

.gx-login-form {
  max-width: 350px;
}

.gx-form-row0 .ant-row {
  margin-left: 0!important;
  margin-right: 0!important;
}

.gx-login-content .ant-form-item-control-wrapper {
  width: 100%;
}

.ant-form-item-control {
  position: relative;
  line-height: 40px;
  zoom: 1;
}

.gx-login-content .ant-form-item-children {
  display: block;
} */