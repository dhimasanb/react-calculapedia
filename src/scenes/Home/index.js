import React, { Component } from "react";
import { Input, Button, Typography, Row, Col } from "antd";
import TableResult from "./components/TableResult";
import { validationAmount } from "../../utils/helpers";
import { fractions } from "../../config/fractions";

class Home extends Component {
  state = {
    amount: "",
    amountLeft: null,
    error: "",
    result: []
  };

  // Event update the react state when user types
  handleChange = name => event => {
    this.setState({ [name]: event.target.value });
  };

  // Keypress event when user press a key 'enter
  handleKeyPressEnter = e => {
    if (e.key === "Enter") {
      this.calculateFractions();
    }
  };

  // Function to calculate fraction rupiah
  calculateFractions = () => {
    let { amount } = this.state;
    const temp = [];
    const result = [];

    try {
      // Function to validation amount
      const validatedAmount = validationAmount(amount);

      amount = validatedAmount;

      // Calculate the fraction of rupiah currency
      while (amount >= 50) {
        for (let i = 0; i < fractions.length; i += 1) {
          if (amount >= fractions[i]) {
            temp[fractions[i]] = temp[fractions[i]]
              ? temp[fractions[i]] + 1
              : 1;
            amount -= fractions[i];
            break;
          }
        }
      }

      // Result sorting by descending
      let resultByOrder = temp.length - 1;
      temp.forEach((value, key) => {
        result[resultByOrder] = { rupiah: key, quantity: value };
        resultByOrder -= 1;
      });

      return this.setState({ amountLeft: amount, result });
    } catch (err) {
      this.setState({ error: err.message });
      return false;
    }
  };

  render() {
    // Object destructuring
    const { amount, result, amountLeft } = this.state;

    return (
      <div style={{ textAlign: "center", paddingTop: "calc(50vh - 166px)" }}>
        <Row>
          <Col span={24}>
            <Typography.Title>Calculapedia</Typography.Title>
          </Col>
        </Row>
        <Row>
          <Col span={20}>
            <Input
              id="input-amount"
              placeholder="Please input amount of money :)"
              value={amount}
              onChange={this.handleChange("amount")}
              onKeyPress={this.handleKeyPressEnter}
            />
          </Col>
          <Col span={4}>
            <Button
              id="btn-calculate"
              type="primary"
              disabled={amount === ""}
              onClick={() => this.calculateFractions()}
              icon="calculator"
            >
              Calculate
            </Button>
          </Col>
        </Row>

        <TableResult data={result} amountLeft={amountLeft} />
      </div>
    );
  }
}

export default Home;
