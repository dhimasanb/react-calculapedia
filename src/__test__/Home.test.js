import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Home from "../scenes/Home";

configure({ adapter: new Adapter() });

describe("Home Scene", () => {
  const wrapper = shallow(<Home />);

  // Home State
  function updateState(amount) {
    it(`updates the amount ${amount} in state`, () => {
      expect(wrapper.state().amount).toEqual(amount);
    });
  }

  function resultState(result) {
    it(`result is ${result}`, () => {
      expect(wrapper.state().result).toEqual(expect.arrayContaining(result));
    });
  }

  function amountLeftState(amountLeft) {
    it(`amount left is ${amountLeft}`, () => {
      expect(wrapper.state().amountLeft).toBe(amountLeft);
    });
  }

  function errorState(error) {
    it(`error ${error}`, () => {
      expect(wrapper.state().error).toBe(error);
    });
  }

  function errorMessageState(errorMessage) {
    it(`error message is ${errorMessage}`, () => {
      expect(wrapper.state().errorMessage).toEqual(errorMessage);
    });
  }

  // Event
  function inputAmount(amount) {
    beforeEach(() => {
      wrapper.find("#input-amount").simulate("change", {
        target: { value: amount }
      });
    });
  }

  function btnCalculateClick() {
    beforeEach(() => wrapper.find("#btn-calculate").simulate("click"));
  }

  function enterKey() {
    beforeEach(() =>
      wrapper.find("#input-amount").simulate("keypress", { key: "Enter" })
    );
  }

  // Valid input function
  function validInput(amount, result, amountLeft) {
    inputAmount(amount);
    updateState(amount);

    describe("click button calculate", () => {
      btnCalculateClick();
      errorState(false);
      resultState(result);
      amountLeftState(amountLeft);
    });

    describe("press enter key", () => {
      enterKey();
      errorState(false);
      resultState(result);
      amountLeftState(amountLeft);
    });
  }

  // Invalid input function
  function invalidInput(amount, error, errorMessage) {
    inputAmount(amount);
    updateState(amount);

    describe("click button calculate", () => {
      btnCalculateClick();
      errorState(error);
      errorMessageState(errorMessage);
    });

    describe("press enter key", () => {
      enterKey();
      errorState(error);
      errorMessageState(errorMessage);
    });
  }

  // Snapshot test
  it("renders properly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // Initializes state test
  it("initializes data in `state`", () => {
    expect(wrapper.state()).toEqual({
      amount: "",
      amountLeft: null,
      error: false,
      errorMessage: "",
      result: []
    });
  });

  // Valid input test
  describe("valid inputs with their canonical equivalents", () => {
    const inputs = [
      {
        amount: "18.215",
        result: [
          { rupiah: 10000, quantity: 1 },
          { rupiah: 5000, quantity: 1 },
          { rupiah: 1000, quantity: 3 },
          { rupiah: 100, quantity: 2 }
        ],
        amountLeft: 15
      },
      {
        amount: "Rp17500",
        result: [
          { rupiah: 10000, quantity: 1 },
          { rupiah: 5000, quantity: 1 },
          { rupiah: 1000, quantity: 2 },
          { rupiah: 500, quantity: 1 }
        ],
        amountLeft: 0
      },
      {
        amount: "Rp17.500,00",
        result: [
          { rupiah: 10000, quantity: 1 },
          { rupiah: 5000, quantity: 1 },
          { rupiah: 1000, quantity: 2 },
          { rupiah: 500, quantity: 1 }
        ],
        amountLeft: 0
      },
      {
        amount: "Rp 120.325",
        result: [
          { rupiah: 100000, quantity: 1 },
          { rupiah: 20000, quantity: 1 },
          { rupiah: 100, quantity: 3 }
        ],
        amountLeft: 25
      },
      {
        amount: "005.000",
        result: [{ rupiah: 5000, quantity: 1 }],
        amountLeft: 0
      },
      {
        amount: "001000",
        result: [{ rupiah: 1000, quantity: 1 }],
        amountLeft: 0
      }
    ];

    inputs.forEach(input => {
      describe(`input: ${input.amount}`, () => {
        validInput(input.amount, input.result, input.amountLeft);
      });
    });
  });

  // Invalid input teest
  describe("invalid inputs", () => {
    const inputs = [
      {
        amount: "17,500",
        error: true,
        message: "Invalid separator!"
      },
      {
        amount: "2 500",
        error: true,
        message: "Invalid separator!"
      },
      {
        amount: "3000 Rp",
        error: true,
        message: "Valid character in wrong position"
      },
      {
        amount: "3000 Rp",
        error: true,
        message: "Valid character in wrong position"
      },
      {
        amount: "Rp",
        error: true,
        message: "Missing value"
      }
    ];

    inputs.forEach(input => {
      describe(`input: ${input.amount}`, () => {
        invalidInput(input.amount, input.error, input.message);
      });
    });
  });
});
