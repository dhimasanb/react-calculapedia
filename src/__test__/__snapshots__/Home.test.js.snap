// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Scene renders properly 1`] = `ShallowWrapper {}`;
